src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.Transcoding.cs

```diff
@@ -0,0 +1,1452 @@
+                    pOutputBufferRemaining = pOutputBuffer;
+                    return OperationStatus.Done;
+                }
+
+                inputLength -= (int)numElementsConverted;
+                outputBytesRemaining -= (int)numElementsConverted;
+            }
+
+            if (inputLength < CharsPerDWord)
+            {
+                goto ProcessInputOfLessThanDWordSize;
+            }
+
+            char* pFinalPosWhereCanReadDWordFromInputBuffer = pInputBuffer + (uint)inputLength - CharsPerDWord;
+
+            // Begin the main loop.
+
+#if DEBUG
+            char* pLastBufferPosProcessed = null; // used for invariant checking in debug builds
+#endif
+
+            uint thisDWord;
+
+            while (pInputBuffer <= pFinalPosWhereCanReadDWordFromInputBuffer)
+            {
+                // Read 32 bits at a time. This is enough to hold any possible UTF16-encoded scalar.
+
+                thisDWord = Unsafe.ReadUnaligned<uint>(pInputBuffer);
+
+            AfterReadDWord:
+
+#if DEBUG
+                Debug.Assert(pLastBufferPosProcessed < pInputBuffer, "Algorithm should've made forward progress since last read.");
+                pLastBufferPosProcessed = pInputBuffer;
+#endif
+
+                // First, check for the common case of all-ASCII chars.
+
+                if (Utf16Utility.AllCharsInUInt32AreAscii(thisDWord))
+                {
+                    // We read an all-ASCII sequence (2 chars).
+
+                    if (outputBytesRemaining < 2)
+                    {
+                        goto ProcessOneCharFromCurrentDWordAndFinish; // running out of space, but may be able to write some data
+                    }
+
+                    // The high WORD of the local declared below might be populated with garbage
+                    // as a result of our shifts below, but that's ok since we're only going to
+                    // write the low WORD.
+                    //
+                    // [ 00000000 0bbbbbbb | 00000000 0aaaaaaa ] -> [ 00000000 0bbbbbbb | 0bbbbbbb 0aaaaaaa ]
+                    // (Same logic works regardless of endianness.)
+                    uint valueToWrite = thisDWord | (thisDWord >> 8);
+
+                    Unsafe.WriteUnaligned<ushort>(pOutputBuffer, (ushort)valueToWrite);
+
+                    pInputBuffer += 2;
+                    pOutputBuffer += 2;
+                    outputBytesRemaining -= 2;
+
+                    // If we saw a sequence of all ASCII, there's a good chance a significant amount of following data is also ASCII.
+                    // Below is basically unrolled loops with poor man's vectorization.
+
+                    uint inputCharsRemaining = (uint)(pFinalPosWhereCanReadDWordFromInputBuffer - pInputBuffer) + 2;
+                    uint minElementsRemaining = (uint)Math.Min(inputCharsRemaining, outputBytesRemaining);
+
+                    if (BitConverter.IsLittleEndian && Bmi2.X64.IsSupported)
```

Comment:
Bmi2 is only available on little-endian. 
```suggestion
                    if (Bmi2.X64.IsSupported)
```